src/Decompiler/Evaluation/ExpressionSimplifier.cs

```diff
@@ -329,7 +329,11 @@ public virtual Expression VisitCast(Cast cast)
             if (exp == Constant.Invalid)
                 return exp;
 
-            var ptCast = cast.DataType as PrimitiveType;
+            var dtCast = cast.DataType;
+            var typeref = dtCast as TypeReference;
+            if (typeref != null)
+                dtCast = typeref.Referent;
+            var ptCast = dtCast as PrimitiveType;
```

Comment:
I'm sorry, I'm confusing `TypeReference` with `TypeVariable` -- again. I need to come up with better names for those classes or find a way to unify them. Pulling now....